### ISSUE TITLE
Represent and manipulate IP addresses without providing a port (Socket::IPAddress without a port)

### ISSUE THREAD

**Issue by j8r** [crystal-lang/crystal#10428]
Sometimes we just want to represent an IP address (https://github.com/crystal-lang/crystal/pull/10408 reminded me of this issue). This has already caught me in the past to have to put a `0` in a place we just want to represent an IP, where we just want to calculate masks for instance.

Examples of other places which has such type:
- [IPAddress ruby gem](https://rubydoc.info/github/bluemonk/ipaddress/master/frames)
- Rust [std::net::IpAddr](https://doc.rust-lang.org/std/net/enum.IpAddr.html)
- Go [net.IP](https://golang.org/pkg/net/#IP)
- C# [system.net.ipaddress](https://docs.microsoft.com/en-us/dotnet/api/system.net.ipaddress?view=net-5.0)

**Comment by j8r** [comment]
Go also has [TCPAddr](https://golang.org/pkg/net/#TCPAddr and [UDPAddr](https://golang.org/pkg/net/#UDPAddr) types, which have a port field.

**Comment by j8r** [comment]
Go also has [TCPAddr](https://golang.org/pkg/net/#TCPAddr) and [UDPAddr](https://golang.org/pkg/net/#UDPAddr) types, which have a port field.

**Comment by straight-shoota** [comment]
Sure, that could be useful sometimes. But IMO it would not be worth it adding another type, when you can just use `Socket::IPAddress` with port 0 for example (and ignore port if you only care about the address).

The naming can obviously be confusing. The type is called `IPAddress` but actually, it consists of the combination of address and port number. This could probably more correctly be called "port", but that would be confusing for similar reasons.

For all practical purposes, I think we're fine with the current situation. It's not ideal, but there's no strictly better solution. And I don't imagine adding another type would be useful for stdlib.

**Comment by straight-shoota** [comment]
When listing other implementations of such a type, you should also mention if there's also a separate type for the combination with a port number like in Go. If not, our `IPAddress` is just like the other languages' type, just with the possibility to also specify a port number.

**Comment by j8r** [comment]
Rust has [SocketAddr](https://doc.rust-lang.org/std/net/enum.SocketAddr.html). C# has [IPEndPoint](https://docs.microsoft.com/en-us/dotnet/api/system.net.ipendpoint?view=net-5).  I don't see anything in Ruby, except `Socket`.

**Comment by j8r** [comment]
Yes, indeed calling the object IPAdress is confusing. It doesn't fit [its definition
](https://en.m.wikipedia.org/wiki/IP_address).

I think it would make perfect sense to have a `SocketAddress`, with a `IPAddress` and a `port` field.

The `0` port is a trick, "a la null pointer". At least, the port should be nillable.
